prov/gni/src/gnix_mr_notifier.c

```diff
@@ -55,69 +57,69 @@ notifier_verify_stuff(struct gnix_mr_notifier *mrn) {
 }
 
 int
-_gnix_notifier_init(struct gnix_mr_notifier *mrn)
+_gnix_notifier_init(void)
 {
-	if (mrn == NULL) {
-		GNIX_INFO(FI_LOG_MR, "mr notifier NULL\n");
-		return -FI_EINVAL;
-	}
-
-	mrn->fd = 0;
-	mrn->cntr = NULL;
-	fastlock_init(&mrn->lock);
+	global_mr_not.fd = -1;
+	global_mr_not.cntr = NULL;
+	fastlock_init(&global_mr_not.lock);
 
 	return FI_SUCCESS;
 }
 
 int
-_gnix_notifier_open(struct gnix_mr_notifier *mrn)
+_gnix_notifier_open(struct gnix_mr_notifier **mrn)
 {
 	int ret = FI_SUCCESS;
 	int kdreg_fd, ret_errno;
         kdreg_get_user_delta_args_t get_user_delta_args;
 
-	if ((mrn->fd != 0) || (mrn->cntr != NULL)) {
-		GNIX_INFO(FI_LOG_MR, "mr notifier already open\n");
-		return -FI_EBUSY;
-	}
+	fastlock_acquire(&global_mr_not.lock);
 
-	fastlock_acquire(&mrn->lock);
+	if (global_mr_not.fd < 0) {
+		kdreg_fd = open(KDREG_DEV, O_RDWR | O_NONBLOCK);
+		if (kdreg_fd < 0) {
+			ret_errno = errno;
+			if (ret_errno != FI_EBUSY) {
+				GNIX_INFO(FI_LOG_MR,
+					  "kdreg device open failed: %s\n",
+					  strerror(ret_errno));
+			}
+			/* Not all of these map to fi_errno values */
+			ret = -ret_errno;
+			goto err_exit;
+		}
 
-	kdreg_fd = open(KDREG_DEV, O_RDWR | O_NONBLOCK);
-	if (kdreg_fd < 0) {
-		ret_errno = errno;
-		if (ret_errno != FI_EBUSY) {
-			GNIX_INFO(FI_LOG_MR, "kdreg device open failed: %s\n",
+		memset(&get_user_delta_args, 0, sizeof(get_user_delta_args));
+		if (ioctl(kdreg_fd, KDREG_IOC_GET_USER_DELTA,
+			  &get_user_delta_args) < 0) {
+			ret_errno = errno;
+			GNIX_INFO(FI_LOG_MR,
+				  "kdreg get_user_delta failed: %s\n",
 				  strerror(ret_errno));
+			close(kdreg_fd);
+			/* Not all of these map to fi_errno values */
+			ret = -ret_errno;
+			goto err_exit;
 		}
-		// Not all of these map to fi_errno values
-		ret = -ret_errno;
-		goto err_exit;
-	}
 
-	(void) memset(&get_user_delta_args,0,sizeof(get_user_delta_args));
-	if (ioctl(kdreg_fd, KDREG_IOC_GET_USER_DELTA,
-		  &get_user_delta_args) < 0) {
-		ret_errno = errno;
-		GNIX_INFO(FI_LOG_MR, "kdreg get_user_delta failed: %s\n",
-			  strerror(ret_errno));
-		close(kdreg_fd);
-		// Not all of these map to fi_errno values
-		ret = -ret_errno;
-		goto err_exit;
-	}
+		if (get_user_delta_args.user_delta == NULL) {
+			GNIX_INFO(FI_LOG_MR, "kdreg get_user_delta is NULL\n");
+			ret = -FI_ENODATA;
+			goto err_exit;
+		}
 
-	if (get_user_delta_args.user_delta == NULL) {
-		GNIX_INFO(FI_LOG_MR, "kdreg get_user_delta is NULL\n");
-		ret = -FI_ENODATA;
-		goto err_exit;
+		global_mr_not.fd = kdreg_fd;
+		global_mr_not.cntr = (kdreg_user_delta_t *)
+				get_user_delta_args.user_delta;
+		global_mr_not.ref_cnt = 1;
+	} else {
```

Comment:
if the ref_cnt was initialized during init, then both paths could simply do ref_cnt++;